design/design-doc-finalizer-support.md

```diff
@@ -0,0 +1,332 @@
+all the Composite resources, and then the Composite resources are deleted from etcd from
+the "bottom up" by garbage collection using the foregroundDeletion finalizer.
+
+As a result of the above:
+- The _--cascade=foreground_ option to kubectl is effectively non-functional on all Crossplane resources.
+- Deletion of the claim or composite returns immediately, while actual managed resource deletion may take a considerable amount of time.
+- Failed deletion of managed resources is not visible to the user. [1612]
+- There is no method available to control the resource deletion process, resulting in orphaned resources as described above.
+
+### Goals
+
+We would like to come up with a design that:
+
+- Will not break any existing functionality.
+- Will support the use of Foreground Cascading Deletion on Composite objects
+- Will support a simulated Foreground Cascading Deletion on Claim objects
+- Facilitates consistent delete behavior for Crossplane resource when compared with native Kubernetes resources
+- Will support "external" finalizers representing declared dependencies on Crossplane resources.
+- Lays the groundwork for supporting ordered deletion of resources
+
+We would like Crossplane to manage resource deletion using the same basic logic as the main
+Kubernetes garbage collection algorithm.  Just as existence of ANY finalizer on a native Kubernetes
+object prevents garbage collection from removing an object from etcd,
+leaving the object "intact" while it is marked as deleted, the existence of an "additional"
+finalizer on a Crossplane resource should block delete routines that would normally
+execute for a resource that has been marked deleted.
+
+Given that a Claim, Composite or Managed Resource in Crossplane is either a collection of other
+Crossplane resources or a reference to an external resource, any dependency represented by the
+existence of a finalizer is actually a dependency on ALL resources owned or represented by that
+resource.  This includes "local" Composite or Managed resources as well as "remote" resources that
+a Managed resource represents.
+
+## Out of Scope
+
+This design does not consider any changes to the Claim/Composite reference design that would
+allow __kubectl delete --cascade=foreground__ to work from the command line on a Claim resource.
+
+This design is not intended to block resource deletion requests to the Kubernetes API.
+
+This design does not address use of the Orphan propagation policy when deleting Claims or
+standalone Composite objects.
+
+## Design
+
+The intent of this design is to allow Crossplane objects to "behave" like native Kubernetes objects
+during delete processing.  A native Kubernetes object will remain in etcd, and all of it's
+content remains available to the user, while the object is marked deleted and waiting for
+finalizers to be removed by any controllers that are watching the resource.
```

Comment:
I think the question here is whether the resource(s) that are associated with the Composite/Managed resource should remain in existence until the `foregroundDeletion` finalizer is removed.   For Composites that just means the `ConnectionDetails` which has little impact. For Managed resources,  if there is an `ownerReference` with `blockOwnerDeletion: true` pointing at it, that Managed Resource is no longer a leaf node in the graph and will have a `foregroundDeletion` finalizer added to it when foreground cascading deletion is requested.  In the current design that finalizer is ignored and the external Delete() function is called immediately when the `deletionTimestamp` is added to the object, causing the external resource to be deleted before the object that is pointing at it can be deleted.  If that external object was relying on the external resource for something it will no longer be available.